stm32: ethernet: mac address: Deprecate Kconfig symbols and generate iud based address by default

erwango opened the pull request:
First, deprecate Kconfig based mac address configuration in profit of existing available device tree properties:
- local-mac-address
- zephyr,random-mac-address

Then, in case user doesn't select any of above options, generate a mac address using device's unique id.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/50947

erwango committed:
drivers: ethernet: stm32_hal: Centralize mac address configuration

Get the mac address be set in a central function rather than split
between 2 places.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers: ethernet: stm32_hal: Move mac address configuration to dt

Deprecate Kconfig mac address related symbols and replace by existing
device tree properties:
- local-mac-address
- zephyr,random-mac-address

User who wants to keep using deprecating method of defining random
MAC address using Kconfig should set CONFIG_ETH_STM32_HAL_RANDOM_MAC=y.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers: stm32: stm32_hal: Generate device id based mac address

In case user didn't set mac address in any way (locally defined or
randomly generated), define mac address using device unique id.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

gautierg-st reviewed (changes requested):
Except typos, LGTM

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -38,6 +38,21 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#endif
+
+#if defined(CONFIG_ETH_STM32_HAL_RANDOM_MAC)
+#warning CONFIG_ETH_STM32_HAL_RANDOM_MAC is deprecated, use device tree "zephyr,random-mac-aadress"
```
Fixed, txs

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1012,15 +1027,23 @@ void HAL_ETH_RxCpltCallback(ETH_HandleTypeDef *heth_handle)
 {
-#if defined(CONFIG_ETH_STM32_HAL_RANDOM_MAC)
+#if defined(ETH_STM32_RANDOM_MAC)
+	/* Either CONFIG_ETH_STM32_HAL_RANDOM_MAC or devie tree property */
```
Fixed, txs

nordicjm commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -54,6 +55,7 @@ config ETH_STM32_HAL_RANDOM_MAC
 	default y
```
If this is deprecated, should select DEPRECATED and also not be default y

nordicjm replied:
And add ` (DEPRECATED)` or similar to end of the text too e.g. `bool "Random MAC address (DEPRECATED)"`

erwango replied:
Ok will update

nordicjm commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -38,6 +39,21 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#if CONFIG_ETH_STM32_HAL_MAC3 || \
+	CONFIG_ETH_STM32_HAL_MAC4 || \
+	CONFIG_ETH_STM32_HAL_MAC5
+#warning CONFIG_ETH_STM32_HAL_MAC* is deprecated, use device tree "local-mac-address"
```
This is not how deprecation is done - see https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated

nordicjm reviewed (changes requested):
Also release note changes needed for deprecation

erwango committed:
doc: release-notes-3.3: Document STM32 ethernet mac address config change

Document changes related to STM32 mac address configuration.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

nordicjm commented on doc/releases/release-notes-3.3.rst:
```diff
@@ -111,6 +111,11 @@ Deprecated in this release
   NOTE: Only functions are marked as ``__deprecated``, type definitions are not.
 
+* STM32 Ethernet Mac address Kconfig related symbols (kconfig:option:`CONFIG_ETH_STM32_HAL_RANDOM_MAC`),
+  kconfig:option:CONFIG_ETH_STM32_HAL_MAC4, ...) have been deprecated in favor
```
```suggestion
  kconfig:option:`CONFIG_ETH_STM32_HAL_MAC4`, ...) have been deprecated in favor
```

erwango commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -48,35 +49,42 @@ config ETH_STM32_HAL_PHY_ADDRESS
 	  The phy address to use.
 
-config ETH_STM32_HAL_RANDOM_MAC
+config ETH_STM32_HAL_RANDOM_MAC (DEPRECATED)
 	bool "Random MAC address"
```
:-), txs

nordicjm commented on doc/releases/release-notes-3.3.rst:
```diff
@@ -350,6 +358,8 @@ Devicetree
 
       * :dtcompatible: `st,stm32-lse-clock`: new ``lse-bypass`` property
+      * :dtcompatible: `st,stm32-ethernet`: now allows ``local-mac-address`` and
+        `zephyr,random-mac-address`` properties.
```
missing 2x quote

nordicjm commented on doc/releases/release-notes-3.3.rst:
```diff
@@ -111,6 +111,11 @@ Deprecated in this release
 
+* STM32 Ethernet Mac address Kconfig related symbols (kconfig:option:`CONFIG_ETH_STM32_HAL_RANDOM_MAC`,
+  kconfig:option:`CONFIG_ETH_STM32_HAL_MAC4`, ...) have been deprecated in favor
+  of the use of zephyr generic device tree `local-mac-address`` and `zephyr,random-mac-address``
```
missing 2x quote on both

erwango replied:
I'm not seeing double quotes on other Kconfig symbols in this file, but I agree there should be a discrepancy somewhere. Anyway: fixed.

nordicjm reviewed (changes requested):
final change!

nordicjm commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -49,34 +50,41 @@ config ETH_STM32_HAL_PHY_ADDRESS
 config ETH_STM32_HAL_RANDOM_MAC
-	bool "Random MAC address"
+	bool "Random MAC address (DEPRECATED)"
 	depends on ENTROPY_GENERATOR
-	default y
+	select DEPRECATED
 	help
 	  Generate a random MAC address dynamically.
+	  Deprecated in favor of device tree property "zephyr,random-mac-address"
 
 if !ETH_STM32_HAL_RANDOM_MAC
 
 config ETH_STM32_HAL_MAC3
-	hex "MAC Address Byte 3"
+	hex "MAC Address Byte 3 (DEPRECATED)"
 	default 0
 	range 0 0xff
+	select DEPRECATED
 	help
 	  This is the byte 3 of the MAC address.
+	  Deprecated in favor of device tree property "local-mac-address"
```
Actually yes, this would always generate a deprecated warning, something doesn't seem right here, maybe use a choice

erwango replied:
@nordicjm 
Also note that `select DEPRECATED` can't be used with these hex symbols:
```
warning: the hex symbol ETH_STM32_HAL_MAC3 (defined at drivers/ethernet/Kconfig.stm32_hal:62) has selects or implies
```
In the end, I don't see a simple change that would not add a new symbol, nor break user configuration.

What about the initial, efficient and simple:
```
#if CONFIG_ETH_STM32_HAL_MAC3 || \
	CONFIG_ETH_STM32_HAL_MAC4 || \
	CONFIG_ETH_STM32_HAL_MAC5
#warning CONFIG_ETH_STM32_HAL_MAC* is deprecated, use device tree "local-mac-address"
```
?

erwango replied:
@nordicjm I reverted to the warning solution. Please tell me if you have better idea.

nordicjm commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -49,34 +50,38 @@ config ETH_STM32_HAL_PHY_ADDRESS
 config ETH_STM32_HAL_RANDOM_MAC
-	bool "Random MAC address"
+	bool "Random MAC address (DEPRECATED)"
 	depends on ENTROPY_GENERATOR
-	default y
+	select DEPRECATED
 	help
 	  Generate a random MAC address dynamically.
+	  Deprecated in favor of device tree property "zephyr,random-mac-address"
 
 if !ETH_STM32_HAL_RANDOM_MAC
 
 config ETH_STM32_HAL_MAC3
-	hex "MAC Address Byte 3"
+	hex "MAC Address Byte 3 (DEPRECATED)"
 	default 0
 	range 0 0xff
 	help
 	  This is the byte 3 of the MAC address.
+	  Deprecated in favor of device tree property "local-mac-address"
 
```
The problem with this is there this file shows 2 possible options, the deprecated random mac, or the 3 byte mac address, it does not allow for a third option. The 3 bytes should be in an if() that only shows if that mode is selected, the 3 modes are random (deprecated), static (deprecated) and dts - so the first option here can be made a choice with those 3 options where dts is the default

erwango replied:
Problem is that current status (before changes) is:
- Default config: Static address used (using default 0)
- If selected: random defined

Adding a choice, user would have to select the new symbol 'static kconfig' to be able to use the deprecated `ETH_STM32_HAL_MACX` symbols. So modify his configuration to be able use a deprecated configuration....
To me the point of deprecation, is to let user keeping his configuration w/o doing nothing, but be informed this wll have to be changed.
If he's forced to change his configuration anyway, no point of using the deprecated symbols..

erwango replied:
@nordicjm Ok, I tried to follow your advise. Please have a new look.

nordicjm commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -38,6 +39,15 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 
+#if CONFIG_ETH_STM32_HAL_MAC3 || CONFIG_ETH_STM32_HAL_MAC4 || CONFIG_ETH_STM32_HAL_MAC5
+#define ETH_STM32_KCONFIG_MAC
+#warning CONFIG_ETH_STM32_HAL_MAC* is deprecated, use device tree "local-mac-address"
```
This line is not needed with the choice, cmake will emit a warning if either of the 2 deprecated choices are selected

erwango commented on doc/releases/release-notes-3.3.rst:
```diff
@@ -129,6 +129,11 @@ Deprecated in this release
 
   NOTE: Only functions are marked as ``__deprecated``, type definitions are not.
 
+* STM32 Ethernet Mac address Kconfig related symbols (kconfig:option:``CONFIG_ETH_STM32_HAL_RANDOM_MAC``,
+  kconfig:option:``CONFIG_ETH_STM32_HAL_MAC4``, ...) have been deprecated in favor
```
got it. fixed

erwango commented on doc/releases/release-notes-3.3.rst:
```diff
@@ -129,6 +129,11 @@ Deprecated in this release
 
+* STM32 Ethernet Mac address Kconfig related symbols (kconfig:option:``CONFIG_ETH_STM32_HAL_RANDOM_MAC``,
+  kconfig:option:``CONFIG_ETH_STM32_HAL_MAC4``, ...) have been deprecated in favor
+  of the use of zephyr generic device tree `local-mac-address`` and `zephyr,random-mac-address``
```
got it. fixed

erwango committed:
drivers: ethernet: stm32_hal: Make MAC address setting method a K choice

In order to ease deprecation of these methods (in favor of dt based
configuration), introduce a choice for MAC address configuration method.

Default to random method (no change).

User wanting to still use ETH_STM32_HAL_MACX symbols should define
CONFIG_ETH_STM32_HAL_USER_STATIC_MAC=y

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

nordicjm reviewed (dismissed):
Not tested it but looks good! Thanks

erwango commented:
@nordicjm Can you have a new look?
I revisited the choice definition to make it optional, which prevents the need to add dummy choices. This is cleaner.

erwango commented:
@ABOSTM FYI

nordicjm reviewed (commented):
Would combine these 2 commits:

- drivers: ethernet: stm32_hal: Move mac address configuration to dt 
- drivers: stm32: stm32_hal: Generate device id based mac address

Reason is to prevent breakage if someone does a bisect, if they end up on the top commit there, networking will not fail to build but will fail to work.

carlescufi commented:
> Would combine these 2 commits:
> 
> * drivers: ethernet: stm32_hal: Move mac address configuration to dt
> * drivers: stm32: stm32_hal: Generate device id based mac address
> 
> Reason is to prevent breakage if someone does a bisect, if they end up on the top commit there, networking will not fail to build but will fail to work.

Given that we do not ensure breakage when bisecting in general, although you are right I will merge this to expedite the process.
